From: Jihong Min <hurryman2212@gmail.com>
Date: Mon, 1 Jun 2026 13:30:00 +0900
Subject: [PATCH] netfilter: flowtable: bind XFRM SOE flows

Teach nft flow offload routing to preserve XFRM packet-offload
destinations and to reclassify SOE-decrypted packets whose secpath
remains attached. This lets the Airoha PPE bind flows that pass through
SOE ESP encapsulation or decapsulation.

Assisted-by: Codex:gpt-5.5 <codex@openai.invalid>
Signed-off-by: Jihong Min <hurryman2212@gmail.com>
---
diff --git a/net/netfilter/nft_flow_offload.c b/net/netfilter/nft_flow_offload.c
index 996a19f..291fc99 100644
--- a/net/netfilter/nft_flow_offload.c
+++ b/net/netfilter/nft_flow_offload.c
@@ -14,6 +14,7 @@
 #include <net/flow.h>
 #include <net/route.h>
 #include <net/ip6_route.h>
+#include <net/xfrm.h>
 #include <net/netfilter/nf_tables.h>
 #include <net/netfilter/nf_tables_core.h>
 #include <net/netfilter/nf_conntrack_core.h>
@@ -34,6 +35,21 @@ static enum flow_offload_xmit_type nft_xmit_type(struct dst_entry *dst)
 	return FLOW_OFFLOAD_XMIT_NEIGH;
 }
 
+static bool nft_flow_offload_xfrm_done(struct sk_buff *skb)
+{
+	struct xfrm_offload *xo;
+	struct sec_path *sp;
+
+	/* SOE returns decrypted packets with secpath still attached. */
+	sp = skb_sec_path(skb);
+	if (!sp)
+		return false;
+
+	xo = xfrm_offload(skb);
+
+	return xo && (xo->flags & CRYPTO_DONE) && (xo->status & CRYPTO_SUCCESS);
+}
+
 static void nft_default_forward_path(struct nf_flow_route *route,
 				     struct dst_entry *dst_cache,
 				     enum ip_conntrack_dir dir)
@@ -416,6 +431,9 @@ static int nft_flow_route_routing(const struct nft_pktinfo *pkt,
 {
 	struct dst_entry *this_dst = skb_dst(pkt->skb);
 	struct dst_entry *other_dst = NULL;
+	struct dst_entry *relookup_dst = NULL;
+	bool this_dst_held = false;
+	bool this_dst_relooked = false;
 	struct flowi fl;
 
 	memset(&fl, 0, sizeof(fl));
@@ -441,7 +459,57 @@ static int nft_flow_route_routing(const struct nft_pktinfo *pkt,
 		break;
 	}
 
-	if (!dst_hold_safe(this_dst))
+	if (this_dst && dst_xfrm(this_dst)) {
+		switch (nft_pf(pkt)) {
+		case NFPROTO_IPV4:
+			fl.u.ip4.flowi4_oif = 0;
+			break;
+		case NFPROTO_IPV6:
+			fl.u.ip6.flowi6_oif = 0;
+			break;
+		}
+	}
+
+	if (nft_flow_offload_xfrm_done(pkt->skb)) {
+		struct flowi cur_fl;
+
+		memset(&cur_fl, 0, sizeof(cur_fl));
+		switch (nft_pf(pkt)) {
+		case NFPROTO_IPV4:
+			cur_fl.u.ip4.daddr = ct->tuplehash[dir].tuple.dst.u3.ip;
+			cur_fl.u.ip4.saddr = ct->tuplehash[dir].tuple.src.u3.ip;
+			cur_fl.u.ip4.flowi4_iif = nft_in(pkt)->ifindex;
+			cur_fl.u.ip4.flowi4_dscp = ip4h_dscp(ip_hdr(pkt->skb));
+			cur_fl.u.ip4.flowi4_mark = pkt->skb->mark;
+			cur_fl.u.ip4.flowi4_flags = FLOWI_FLAG_ANYSRC;
+			break;
+		case NFPROTO_IPV6:
+			cur_fl.u.ip6.daddr =
+				ct->tuplehash[dir].tuple.dst.u3.in6;
+			cur_fl.u.ip6.saddr =
+				ct->tuplehash[dir].tuple.src.u3.in6;
+			cur_fl.u.ip6.flowi6_iif = nft_in(pkt)->ifindex;
+			cur_fl.u.ip6.flowlabel =
+				ip6_flowinfo(ipv6_hdr(pkt->skb));
+			cur_fl.u.ip6.flowi6_mark = pkt->skb->mark;
+			cur_fl.u.ip6.flowi6_flags = FLOWI_FLAG_ANYSRC;
+			break;
+		}
+
+		nf_route(nft_net(pkt), &relookup_dst, &cur_fl, false,
+			 nft_pf(pkt));
+		if (relookup_dst) {
+			this_dst = relookup_dst;
+			this_dst_relooked = true;
+		}
+	}
+
+	if (this_dst_relooked)
+		this_dst_held = true;
+	else if (dst_hold_safe(this_dst))
+		this_dst_held = true;
+
+	if (!this_dst_held)
 		return -ENOENT;
 
 	nf_route(nft_net(pkt), &other_dst, &fl, false, nft_pf(pkt));
@@ -449,6 +517,11 @@ static int nft_flow_route_routing(const struct nft_pktinfo *pkt,
 		dst_release(this_dst);
 		return -ENOENT;
 	}
+	other_dst = xfrm_lookup(nft_net(pkt), other_dst, &fl, NULL, 0);
+	if (IS_ERR(other_dst)) {
+		dst_release(this_dst);
+		return -ENOENT;
+	}
 
 	nft_default_forward_path(route, this_dst, dir);
 	nft_default_forward_path(route, other_dst, !dir);
@@ -556,9 +629,24 @@ static void nft_flow_set_dscp(const struct nft_pktinfo *pkt,
 
 static bool nft_flow_offload_skip(struct sk_buff *skb, int family)
 {
-	if (skb_sec_path(skb))
+	struct xfrm_offload *xo;
+	struct sec_path *sp;
+
+	sp = skb_sec_path(skb);
+	if (sp) {
+		struct xfrm_state *x = sp->len ? sp->xvec[sp->len - 1] : NULL;
+
+		xo = xfrm_offload(skb);
+		if (xo && (xo->flags & CRYPTO_DONE) &&
+		    (xo->status & CRYPTO_SUCCESS))
+			goto out;
+
+		if (!x || x->xso.type != XFRM_DEV_OFFLOAD_PACKET ||
+		    !x->xso.dev || !(x->xso.dev->features & NETIF_F_HW_ESP))
-		return true;
+			return true;
+	}
 
+out:
 	if (family == NFPROTO_IPV4) {
 		const struct ip_options *opt;
 
@@ -590,6 +678,7 @@ static void nft_flow_offload_eval(const struct nft_expr *expr,
 	struct nf_flow_route route = {};
 	enum ip_conntrack_info ctinfo;
 	struct flow_offload *flow;
+	bool tcp_not_established = false;
 	enum ip_conntrack_dir dir;
 	struct nf_conn *ct;
 	int ret;
@@ -605,8 +694,13 @@ static void nft_flow_offload_eval(const struct nft_expr *expr,
 	case IPPROTO_TCP:
 		tcph = skb_header_pointer(pkt->skb, nft_thoff(pkt),
 					  sizeof(_tcph), &_tcph);
-		if (unlikely(!tcph || tcph->fin || tcph->rst ||
-			     !nf_conntrack_tcp_established(ct)))
-			goto out;
+		if (unlikely(!tcph))
+			goto out;
+		if (unlikely(tcph->fin || tcph->rst))
+			goto out;
+		if (unlikely(!nf_conntrack_tcp_established(ct))) {
+			tcp_not_established = true;
+			break;
+		}
 		break;
 	case IPPROTO_UDP:
@@ -654,8 +743,18 @@ static void nft_flow_offload_eval(const struct nft_expr *expr,
 		if (nft_flow_route_routing(pkt, ct, &route, dir,
 					   priv->flowtable) < 0)
 			goto err_flow_route;
 	}
 
+	/* XFRM/SOE packet offload can hide part of the plaintext TCP handshake
+	 * from conntrack: the encrypted/decrypt side may be bound first, then
+	 * the return tuple is learned as XFRM. Keep the normal TCP-established
+	 * requirement for all non-XFRM routes.
+	 */
+	if (tcp_not_established &&
+	    route.tuple[IP_CT_DIR_ORIGINAL].xmit_type != FLOW_OFFLOAD_XMIT_XFRM &&
+	    route.tuple[IP_CT_DIR_REPLY].xmit_type != FLOW_OFFLOAD_XMIT_XFRM)
+		goto err_flow_route;
+
 	flow = flow_offload_alloc(ct);
 	if (!flow)
 		goto err_flow_alloc;
